src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2252,32 +2240,26 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
 		{
 			int shardIndex = shardInterval->shardIndex;
 
-			taskRequiredForShardIndex[shardIndex] = true;
-
+			taskRequiredForShardIndex =
+				bms_add_member(taskRequiredForShardIndex, shardIndex);
 			minShardOffset = Min(minShardOffset, shardIndex);
-			maxShardOffset = Max(maxShardOffset, shardIndex);
 		}
 	}
 
 	/*
-	 * To avoid iterating through all shards indexes we keep the minimum and maximum
-	 * offsets of shards that were not pruned away. This optimisation is primarily
-	 * relevant for queries on range-distributed tables that, due to range filters,
-	 * prune to a small number of adjacent shards.
+	 * We keep track of minShardOffset to skip over a potentially big amount of pruned
+	 * shards. However, we need to start at minShardOffset - 1 to make sure we don't
+	 * miss to first/min shard recorder as bms_next_member will return the first member
+	 * added after shardOffset. Meaning minShardOffset would be the first member we
+	 * expect.
 	 *
-	 * In other cases, such as an OR condition on a hash-distributed table, we may
-	 * still visit most or all shards even if some of them were pruned away. However,
-	 * given that hash-distributed tables typically only have a few shards the
-	 * iteration is still very fast.
+	 * We don't have to keep track of maxShardOffset as the bitmapset will only have been
+	 * allocated till the last shard we have added. Therefore, the iterator will quickly
+	 * identify the end of the bitmapset.
 	 */
-	for (int shardOffset = minShardOffset; shardOffset <= maxShardOffset; shardOffset++)
+	int shardOffset = minShardOffset - 1;
```

Comment:
it seems the usual method is to initialize at -1, would that work here?

Reply:
It would, and it means we could remove `minShardOffset`. The only downside of that would be that on a 100k shardcount we will have 100k/64 (on x86_64) iterations to find the first bit set in the Bitmapset, if the selected shards are at the end of the sequence.

As per commit message and PR description, I have kept the `minShardOffset` to keep the optimization we had for such situations in favour of the readability by keeping track of the lowest added shard during the population.

Reply:
makes sense, thanks